sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala

```diff
@@ -105,18 +105,27 @@ object V2ExpressionUtils extends SQLConfHelper with Logging {
           TransformExpression(bound, resolvedRefs, Some(numBuckets))
         }
       }
-    case NamedTransform(name, refs)
-        if refs.length == 1 && refs.forall(_.isInstanceOf[NamedReference]) =>
```

Comment:
This `if` condition was prohibiting transforms with literals and multiple args.

Reply:
Hmm I remember added this check since storage partitioned join can only handle one named reference in partition transform at the moment. There's an [assertion](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala#L656) in `KeyGroupedShuffleSpec` for this: 

Could you add a test in `EnsureRequirementsSuite` for a `NamedTransform` with >1 named references? just to make sure it's not breaking anything.

Reply:
Let me check.

Reply:
On the other hand, skipping `KeyGroupedPartitioning` means we will never create `KeyGroupedShuffleSpec`, which has that validation logic. Maybe, that's a safer option.

Reply:
+1 to this option. We should improve `KeyGroupedPartitioning.satisfies0` late.

Reply:
I'm also +1 on the first option. I also tried on my side, and it's possible to trigger the assertion error with the following example:
```scala
  test("multiple named references in transform") {
    val plan1 = DummySparkPlan(
      outputPartitioning = KeyGroupedPartitioning(
        truncate(exprD, exprA) :: bucket(16, exprB) :: Nil, 4)
    )
    val plan2 = DummySparkPlan(
      outputPartitioning = KeyGroupedPartitioning(
        truncate(exprD, exprA) :: bucket(16, exprC) :: Nil, 4)
    )
    val smjExec = SortMergeJoinExec(
      exprD :: exprA :: exprB :: Nil, exprD :: exprA :: exprC :: Nil, Inner, None, plan1, plan2)
    EnsureRequirements.apply(smjExec) match {
      case SortMergeJoinExec(_, _, _, _,
        SortExec(_, _, DummySparkPlan(_, _, left: KeyGroupedPartitioning, _, _), _),
        SortExec(_, _, DummySparkPlan(_, _, right: KeyGroupedPartitioning, _, _), _), _) =>
        assert(left.expressions === Seq(truncate(exprD, exprA), bucket(16, exprB)))
        assert(right.expressions === Seq(truncate(exprD, exprA), bucket(16, exprC)))
      case other => fail(other.toString)
    }
  }
```